url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/mxgraph/io/mxCellCodec.java

```diff
@@ -0,0 +1,220 @@
+ * dynamically at load time and used implicitely via mxCodec
+ * and the mxCodecRegistry.
+ */
+public class mxCellCodec extends mxObjectCodec
+{
+
+	/**
+	 * Constructs a new cell codec.
+	 */
+	public mxCellCodec()
+	{
+		this(new mxCell(), null, new String[] { "parent", "source", "target" },
+				null);
+	}
+
+	/**
+	 * Constructs a new cell codec for the given template.
+	 */
+	public mxCellCodec(Object template)
+	{
+		this(template, null, null, null);
+	}
+
+	/**
+	 * Constructs a new cell codec for the given arguments.
+	 */
+	public mxCellCodec(Object template, String[] exclude, String[] idrefs,
+			Map<String, String> mapping)
+	{
+		super(template, exclude, idrefs, mapping);
+	}
+
+	/**
+	 * Excludes user objects that are XML nodes.
+	 */
+	public boolean isExcluded(Object obj, String attr, Object value,
+			boolean write)
+	{
+		return exclude.contains(attr)
+				|| (write && attr.equals("value") && value instanceof Node && ((Node) value)
+						.getNodeType() == Node.ELEMENT_NODE);
+	}
+
+	/**
+	 * Encodes an mxCell and wraps the XML up inside the
+	 * XML of the user object (inversion).
+	 */
+	public Node afterEncode(mxCodec enc, Object obj, Node node)
+	{
+		if (obj instanceof mxCell)
+		{
+			mxCell cell = (mxCell) obj;
+
+			if (cell.getValue() instanceof Node)
+			{
+				// Wraps the graphical annotation up in the
+				// user object (inversion) by putting the
+				// result of the default encoding into
+				// a clone of the user object (node type 1)
+				// and returning this cloned user object.
+				Element tmp = (Element) node;
+				node = enc.getDocument().importNode((Node) cell.getValue(),
+						true);
+				node.appendChild(tmp);
+
+				// Moves the id attribute to the outermost
+				// XML node, namely the node which denotes
+				// the object boundaries in the file.
+				String id = tmp.getAttribute("id");
+				((Element) node).setAttribute("id", id);
+				tmp.removeAttribute("id");
+			}
+		}
+
+		return node;
+	}
+
+	/**
+	 * Decodes an mxCell and uses the enclosing XML node as
+	 * the user object for the cell (inversion).
+	 */
+	public Node beforeDecode(mxCodec dec, Node node, Object obj)
+	{
+		Element inner = (Element) node;
+
+		if (obj instanceof mxCell)
+		{
+			mxCell cell = (mxCell) obj;
+			String classname = getName();
+
+			if (!node.getNodeName().equals(classname))
+			{
+				// Passes the inner graphical annotation node to the
+				// object codec for further processing of the cell.
+				Node tmp = inner.getElementsByTagName(classname).item(0);
+
+				if (tmp != null && tmp.getParentNode() == node)
+				{
+					inner = (Element) tmp;
+
+					// Removes annotation and whitespace from node
+					Node tmp2 = tmp.getPreviousSibling();
+
+					while (tmp2 != null && tmp2.getNodeType() == Node.TEXT_NODE)
+					{
+						Node tmp3 = tmp2.getPreviousSibling();
+
+						if (tmp2.getTextContent().trim().length() == 0)
+						{
+							tmp2.getParentNode().removeChild(tmp2);
+						}
+
+						tmp2 = tmp3;
+					}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 120:130, 135:145.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.